browser/ui/webui/settings/brave_import_bulk_data_handler.cc

```diff
@@ -0,0 +1,174 @@
+                      created_profile->GetPrefs()->SetBoolean(
+                          prefs::kHasSeenWelcomePage, true);
+                      std::move(initialized_callback).Run(created_profile);
+                    },
+                    std::move(profile_callback)));
+          },
+          std::move(profile_ready_callback), name));
+}
+
+void BraveImportBulkDataHandler::HandleImportDataBulk(
+    const base::Value::List& args) {
+  CHECK_GE(args.size(), 2u);
+  const auto& list = args[0].GetList();
+  // Bulk profiles import assumes new profiles will be created on our side if
+  // they do not exist.
+  const base::Value& types = args[1];
+  for (const auto& it : list) {
+    int browser_index = it.GetInt();
+    importing_profiles_.insert(browser_index);
+    base::Value::List single_profile_args;
+    single_profile_args.Append(base::Value(browser_index));
+    single_profile_args.Append(types.Clone());
+    BraveImportDataHandler::HandleImportData(single_profile_args);
+  }
+}
+
+absl::optional<int> BraveImportBulkDataHandler::GetProfileIndex(
+    const importer::SourceProfile& source_profile) {
+  for (auto index : importing_profiles_) {
+    const auto& profile = GetSourceProfileAt(index);
+    if (profile.source_path == source_profile.source_path) {
+      return index;
+    }
+  }
+  return absl::nullopt;
+}
+
+void BraveImportBulkDataHandler::StartImport(
+    const importer::SourceProfile& source_profile,
+    uint16_t imported_items) {
+  // If profile is not from the bulk import request fallback to single profile
+  // import.
+  if (!GetProfileIndex(source_profile).has_value()) {
+    BraveImportDataHandler::StartImport(source_profile, imported_items);
+    return;
+  }
+  if (!imported_items)
+    return;
+  auto profile_name = source_profile.profile.empty()
+                          ? source_profile.importer_name
+                          : source_profile.profile;
+  auto import_callback = base::BindOnce(
+      &BraveImportBulkDataHandler::StartImportImpl, weak_factory_.GetWeakPtr(),
+      source_profile, imported_items);
+#if BUILDFLAG(IS_MAC)
+  CheckDiskAccess(imported_items, source_profile.source_path,
+                  source_profile.importer_type,
+                  base::BindOnce(&BraveImportBulkDataHandler::PrepareProfile,
+                                 weak_factory_.GetWeakPtr(), profile_name,
+                                 std::move(import_callback)));
+#else
+  PrepareProfile(profile_name, std::move(import_callback));
+#endif
+}
+
+void BraveImportBulkDataHandler::NotifyImportProgress(
+    const importer::SourceProfile& source_profile,
+    const base::Value& info) {
+  FireWebUIListener("brave-import-data-status-changed", info);
+}
+
+void BraveImportBulkDataHandler::OnImportEnded(
+    const importer::SourceProfile& source_profile) {
+  auto index = GetProfileIndex(source_profile);
+  if (index.has_value()) {
+    importing_profiles_.erase(index.value());
+  }
```

Comment:
I think we should call `BraveImportDataHander::OnImportEnded()` for non-bulk mode?

Reply:
added